tests/rbd/test_performance_immutable_cache.py

```diff
@@ -0,0 +1,236 @@
+            "operations": {
+                "fs": "ext4",
+                "io": True,
+                "mount": True,
+                "nounmap": True,
+                "device_map": True,
+                "io_type": io_type,
+            },
+            "skip_mkfs": False,
+            "cmd_timeout": 2400,
+        },
+    }
+    krbd_io_handler(**io_config)
+    return io_config["config"].get("time_taken", "")
+
+
+def test_immutable_cache_performance(rbd_obj, **kw):
+    """
+    Test the performance of Immutable Object Cache with and without cache.
+
+    Args:
+        rbd_obj (object): RBD object.
+        kw : test data.
+
+    Raises:
+        Exception: If any test step fails.
+    """
+    config = kw.get("config")
+    ceph_cluster = kw.get("ceph_cluster")
+    client = ceph_cluster.get_nodes(role="client")[0]
+    try:
+        if configure_immutable_cache(rbd_obj, client, "user1"):
+            log.error("Immutable object cache configuration failed")
+            return 1
+        log.info("Immutable object cache configuration completed successfully.")
+        pdb.set_trace()
+        for pool_type in rbd_obj.get("pool_types"):
+            rbd_config = kw.get("config", {}).get(pool_type, {})
+            multi_pool_config = getdict(rbd_config)
+
+            for pool, pool_config in multi_pool_config.items():
+                multi_image_config = getdict(pool_config)
+                for image in multi_image_config.keys():
+                    image_spec = f"{pool}/{image}"
+
+                    config["image_spec"] = image_spec
+                    rbd = rbd_obj.get("rbd")
+
+                    mount_path_with_cache = f"/tmp/mnt_{rbd_obj.random_string(len=5)}"
+                    mount_path_without_cache = (
+                        f"/tmp/mnt_{rbd_obj.random_string(len=5)}"
+                    )
+
+                    # Write data to the image using FIO with cache enabled
+                    time_with_cache = create_and_test_io(
+                        client, rbd, config, mount_path_with_cache, "write"
+                    )
+
+                    # create snapshot, protect and clone it for the image
+                    snap = "image_snap" + random_string(len=5)
+                    clone = "image_clone" + random_string(len=5)
+                    snap_spec = f"{pool}/{image}@{snap}"
+                    clone_spec = f"{pool}/{clone}"
+
+                    create_snap_and_clone(rbd, snap_spec, clone_spec)
+
+                    # Read IO from cloned images to get cache files
+                    time_with_cache += create_and_test_io(
+                        client, rbd, config, mount_path_with_cache, "read"
+                    )
+
+                    # Removing immutable cache
+                    client.exec_command(cmd=f"rm -rf {immutable_cache_path}", sudo=True)
+
+                    # Write data to the image using FIO without cache
+                    time_without_cache = create_and_test_io(
```

Comment:
I remember reading somewhere that immutable cache is basically read-only cache? What would that mean in this context? Does it mean that it will speed up only the reads? If that's not the case, can you check whether it is speeding up writes separately by comparing write time and reads separately by comparing read time?

Reply:
i am facing some issue in performance http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-GAX2Z9/verify_performance_for_Immutable_object_cache_0.log
without cache only it's taking less time.